Updates for the Tutorial and Plans content (#7482)
roc-lang/roc: www/content/tutorial.md

### [The `try` keyword](#the-try-keyword) {#the-try-keyword}

Roc has a `try` keyword, which is convenient syntax sugar for `Result.try`. For example, consider the following `get_letter` function:

```roc
get_letter : Str -> Result Str [OutOfBounds, InvalidNumStr]
get_letter = \index_str ->
    index = try Str.toU64 index_str
    List.get ["a", "b", "c", "d"] index
```

Here's what this does:

- If the `Str.toU64` function returns an `Ok` value, then `try` will return what's inside the `Ok`. For example:
  - If we call `get_letter "2"`, then `Str.toU64` returns `Ok 2`, and the `try` unwraps to the integer 2, so `index` is set to 2 (not `Ok 2`). Then the `List.get` function is called and returns `Ok "c"`.
  - If the `Str.toU64` function returns an `Err` value, then the `try` keyword immediately interrupts the `get_letter` function and makes it return this error.
  - For example, if we call `get_letter "abc"`, then the call to `Str.toU64` returns `Err InvalidNumStr`, and the `try` keyword ensures that the `get_letter` function returns this error immediately, without executing the rest of the function.

> Thanks to the `try` keyword, your code can focus on the "happy path" (where nothing fails) and simply bubble up to the caller any error that might occur. Your error handling code can be neatly separated, and you can rest assured that you won't forget to handle any errors, since the compiler will let you know. See this [code example](https://github.com/roc-lang/examples/blob/main/examples/Results/main.roc) for more details on error handling.

Reviewer: ```suggestion
Thanks to the `try` keyword, your code can focus on the "happy path" (where nothing fails) and simply bubble up to the caller any error that might occur. Your error handling code can be neatly separated, and you can rest assured that you won't forget to handle any errors, since the compiler will let you know. See this [code example](https://www.roc-lang.org/examples/Results/README.html) for more details on error handling.
```

## [Effectful functions](#efffectful-functions) {#efffectful-functions}

There are two types of functions in roc, "pure" and "effectful". Consider these two functions:

```roc
with_extension : Str -> Str
with_extension = \filename ->
    "$(filename).roc"

read_file! : Str => Str
read_file! = \path ->
    File.read_utf8! with_extension(path) |> Result.withDefault ""
```

Notice the subtle difference in these functions' types:

- Str `->` Str is the type of the pure function `with_extension`
- Str `=>` Str is the type of the effectful function `read_file!`

The arrow in the type tells you whether a function is pure or might perform effects. Pure functions use `->` and effectful functions use `=>`. Effectful functions can call either pure functions or other effectful functions, but pure functions can only call other pure functions. (The fact that pure functions can only call pure functions is part of the definition of pure functions; Roc is just reflecting that fact in the type system.)

The `!` is a suffix character inside function names themselves, marking them as effectful. This is just a naming convention, though it is enforced by the compiler with a warning.

> Note: this was a previous change to the language, and `!` used to be an operator, so you may still find code that as a `Task` in it or an effectful function like `main` without a `!` suffix. This code should be upgraded to use `Result` and the naming convention for purity inference, e.g. `main!`. For more information about this change, see [this recent talk](https://www.youtube.com/watch?v=42TUAKhzlRI).

Reviewer: ```suggestion
Note: this was a previous change to the language, and `!` used to be an operator, so you may still find code that has a `Task` in it or an effectful function like `main` without a `!` suffix. This code should be upgraded to use `Result` and the naming convention for purity inference, e.g. `main!`. For more information about this change, see [this recent talk](https://www.youtube.com/watch?v=42TUAKhzlRI).
```

Also:
* I don't fully comprehend what this fragment "this was a previous change to the language" is about. But it may be just me.
* The "recent talk" may shortly become not-so-recent. Perhaps the word "recent" can be removed.

Author: Done in 7cd9fae334

Let's look at some examples. In the `basic-cli` platform, here are four operations we can do:

- Write a string to the terminal
- Read a string from user input
- Write a string to a file
- Read a string from a file

We'll use these four operations to learn about effects.

Let's revisit that first application.

```roc
app [main!] { pf: platform "https://github.com/roc-lang/basic-cli/releases/download/0.18.0/0APbwVN1_p1mJ96tXjaoiUCr8NBGamr8G8Ac_DrXR-o.tar.br" }

import pf.Stdout

main! = \_args ->
    Stdout.line! "Hi there, from inside a Roc app. 🎉"
```

This code prints "Hi there, from inside a Roc app. 🎉" to the [standard output](<https://en.wikipedia.org/wiki/Standard_streams#Standard_output_(stdout)>). `Stdout.line` has this type:

```roc
Stdout.line! : Str => Result {} [StdoutErr IOErr]
```

> An effectulful function is capable of interacting with state outside your Roc program, such as the terminal's standard output, or a file.

Reviewer: ```suggestion
An effectful function is capable of interacting with state outside your Roc program, such as the terminal's standard output, or a file.
```

Note that to call this `Stdin.line!`, you need to provide the unit value `{}` as an argument.

### [Reading values](#reading-values) {#reading-values}

Let's change `main!` to read a line from `stdin`, and then print what we got:

```roc
app [main!] { pf: platform "https://github.com/roc-lang/basic-cli/releases/download/0.18.0/0APbwVN1_p1mJ96tXjaoiUCr8NBGamr8G8Ac_DrXR-o.tar.br" }

import pf.Stdout
import pf.Stdin

main! = \_args ->
    try Stdout.line! "Type in something and press Enter:"
    input = try Stdin.line! {}
    try Stdout.line! "Your input was: $(input)"

    Ok {}
```

If you run this program, it will print "Type in something and press Enter:" and then pause.
That's because it's waiting for you to type something in and press Enter! Once you do,
it should print back out what you entered.

> ### [Effectul failure](#failure) {#failure}

Reviewer: ```suggestion
### [Effectful failure](#failure) {#failure}
```

it with this one:

```roc
main! : _ => Result {} [Exit I32 Str]
main! = \_args -> Err (Exit 42 "An error happened!")
```

Now if we run the application, it will print the line "An error happened!" to stderr and exit with a status code of 42. (You can check the status code of the most recent terminal command that finished in Windows by running `echo %ERRORLEVEL%` (or `$LASTEXITCODE` in PowerShell), or by running `echo $?` in macOS or Linux.)

Now let's try running it with this version of `main`:

```roc
main! : _ => Result {} [Exit I32 Str]
main! = \_args -> Ok {}
```

This program won't print anything at all, but it will exit with a status code of `0`, indicating success.

In summary:

- If the `main!` function ends in a `Ok {}`, then it means the final expression succeeded and the program will exit with status code 0.
- If the `main!` function ends in a `Err (Exit 42 "…")`, then it means it failed, and the only information we got about the failure was that the program should exit with code 42 instead of 0, and that it should print a particular string to stderr to inform the user about what happened.

> ### [Handling failure](#handling-failures) {#handling-failures}

Reviewer: I suggest to either add the final `s` to the title or remove it from the identifier.

Author: Done in 53f5fc3e79

Since `Stdout.line! : Str => Result {} [StdoutErr IOErr]`, in the above we are unwrapping the `Result` returned by calling these effects.

An alternative option, is to just ignore the return value entirely:

```roc
main! = \_args ->
    _ = Stdout.line! "Type in something and press Enter:"
    when Stdin.line! {} is
        Ok input ->
            _ = Stdout.line! "Your input was: $(input)"
            Ok {}
        Err _ ->
            Ok {}
```

### [Tagging errors](#tagging-errors) {#tagging-errors}

Although it's rare, it is possible that either of the `Stdout.line!` operations in our example could fail:

```roc
main! = \_args ->
    try Stdout.line! "Type something and press Enter."
    input = try Stdin.line! {}
    try Stdout.line! "You entered: $(input)"
    Ok {}
```

(In this particular example, it's very unlikely that this would come up at all, and even if it did, we might not care which one caused the problem. But you can imagine having multiple HTTP requests, or file writes, and wanting to know which of them was the one that failed.)

> If an error happended here, we wouldn't know which effectul function was the cause of the failure.

Reviewer: ```suggestion
If an error happended here, we wouldn't know which effectful function was the cause of the failure.
```

One option is to "tag the error" using `Result.mapErr` to wrap the error in a [tag](#tags) like so:

```roc
app [main!] { pf: platform "https://github.com/roc-lang/basic-cli/releases/download/0.18.0/0APbwVN1_p1mJ96tXjaoiUCr8NBGamr8G8Ac_DrXR-o.tar.br" }

import pf.Stdout
import pf.Stdin

main! = \_args ->

    Stdout.line! "Type something and press Enter."
    |> Result.mapErr UnableToPrintPrompt
    |> try

    input =
        Stdin.line! {}
        |> Result.mapErr UnableToReadInput
        |> try

    Stdout.line! "You entered: $(input)"
    |> Result.mapErr UnableToPrintInput
    |> try

    Ok {}
```

The `mapErr` function has this type:

```
Result.mapErr : Result ok a, (a -> b) -> Result ok b
```

> Here we're passing in "tagging functions"—namely, `UnableToPrintPrompt` and `UnableToReadInput`. (See [Using tags as functions](#using-tags-as-functions) for how this works.)

Reviewer: ```suggestion
Here we're passing in "tagging functions" — namely, `UnableToPrintPrompt` and `UnableToReadInput`. (See [Using tags as functions](#using-tags-as-functions) for how this works.)
```
